// @cubejs-backend/shared/dist/src/env.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDockerImage = exports.getEnv = void 0;
const env_var_1 = require("env-var");
const variables = {
    devMode: () => env_var_1.get('CUBEJS_DEV_MODE')
        .default('false')
        .asBoolStrict(),
    port: () => env_var_1.get('PORT')
        .default(4000)
        .required()
        .asPortNumber(),
    tlsPort: () => env_var_1.get('TLS_PORT')
        .default(4433)
        .required()
        .asPortNumber(),
    tls: () => env_var_1.get('CUBEJS_ENABLE_TLS')
        .default('false')
        .asBoolStrict(),
    webSockets: () => env_var_1.get('CUBEJS_WEB_SOCKETS')
        .default('false')
        .asBoolStrict(),
    dockerImageVersion: () => env_var_1.get('CUBEJS_DOCKER_IMAGE_VERSION')
        .asString(),
    // It's only excepted for CI, nothing else.
    internalExceptions: () => env_var_1.get('INTERNAL_EXCEPTIONS_YOU_WILL_BE_FIRED')
        .default('false')
        .asEnum(['exit', 'log', 'false'])
};
function getEnv(key) {
    return variables[key]();
}
exports.getEnv = getEnv;
function isDockerImage() {
    return Boolean(process.env.CUBEJS_DOCKER_IMAGE_TAG);
}
exports.isDockerImage = isDockerImage;
